src/NodeDRListener.cpp

```diff
@@ -99,18 +108,80 @@ void NodeDRListener::reserve(CORBA::ULong)
 
 void NodeDRListener::push_back(const DDS::SampleInfo& src, const void* sample)
 {
+  if (unsubscribing_) {
+    return;
+  }
+
   Local<Value> argv[] = {Nan::New(js_dr_), Handle<Value>(copyToV8(src)),
                          src.valid_data ? conv_.toV8(sample).As<Value>()
                          : Nan::Undefined().As<Value>()};
 
   Local<Function> callback = Nan::New(callback_);
   Nan::Callback cb(callback);
   cb.Call(sizeof(argv) / sizeof(argv[0]), argv);
+}
+
+void NodeDRListener::unsubscribe() {
+  if (receiving_samples_) {
+    // Inform the Listener to skip any remaining takes and use
+    // UnsubscibeWorker to unsubscribe at the next opportunity.
+    unsubscribing_ = true;
+  } else { // Unsubscribe Now
+    unsubscribe_now();
+  }
+}
+
+void NodeDRListener::unsubscribe_now() {
+  Local<v8::Object> dr_js = Nan::New(js_dr_);
+  void* const dr_obj = Nan::GetInternalFieldPointer(dr_js, 0);
+  DDS::DataReader_var dr = static_cast<DDS::DataReader*>(dr_obj);
+  Nan::SetInternalFieldPointer(dr_js, 0, 0);
+
+  _add_ref();
```

Comment:
So If I keep this _add_ref, the Listener doesn't get deleted automatically, otherwise it does. I'm I doing something wrong somewhere or does this line actually have to be removed?

Reply:
I don't think I understand, should the DataReader be deleting the Node DataReaderListener when DataReader_var is set to 0 or at least be decrementing the reference count? It doesn't seem to be doing either and currently how RcObject and TAO's Object vars work together (or don't) is still fuzzy to me.

Reply:
The DataReader has a reference to the listener, that reference is dropped in DataReaderImpl::cleanup().
The V8 object also has a reference to the listener, which is managed as I described above.

Since it uses a reference counting scheme for object lifetime, we don't know exactly when it it will be deleted.  However, unless a reference is leaked or there's a cycle, the reference count eventually goes to 0 causing the object to be deleted.

RcObject is not involved here because the object already has a reference count which is required by the IDL-to-C++ mapping.  RcObject is used for internal OpenDDS objects that don't inherit from an IDL-defined interface.

Reply:
For the record, the notes about RcObject above apply to version 3.12 of OpenDDS.  There are changes on master after 3.12 which we'll have to deal with one way or another.

Reply:
I looked at the reference counts of `NodeDRListener` during the shutdown. I used `ref_count()` to access the `RcObject`, it starts at 2 when `unsubscribe_now` starts, goes to 3 when `_add_ref` is called and goes back to 2 after the subscriber deletes it's entities. The `RcObject` ref count goes to one after `close_cb`. I used `_refcount_value()` to access the TAO reference count, but it is always one. The destructor is never called as is.